core/pta/ocall.c

```diff
@@ -0,0 +1,364 @@
+		case TEE_PARAM_TYPE_VALUE_INPUT:
+			break;
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+		case TEE_PARAM_TYPE_VALUE_OUTPUT:
+			up->vals[n * 2] = params[n].u.value.a;
+			up->vals[n * 2 + 1] = params[n].u.value.b;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+			buffer = (void *)(uintptr_t)up->vals[n * 2];
+			size = up->vals[n * 2 + 1];
+			if (params[n].u.memref.size != size)
+				return TEE_ERROR_BAD_PARAMETERS;
+			if (ADD_OVERFLOW(mobj_offs, size, &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+			buffer = (void *)(uintptr_t)up->vals[n * 2];
+			size = up->vals[n * 2 + 1];
+			if (params[n].u.memref.size > size)
+				return TEE_ERROR_BAD_PARAMETERS;
+			source = PTR_ADD(mobj_va, mobj_offs);
+			memcpy(buffer, source, params[n].u.memref.size);
+			if (ADD_OVERFLOW(mobj_offs, size, &mobj_offs))
+				return TEE_ERROR_SECURITY;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result ocall_send(struct tee_ta_session *session,
+			     struct tee_ta_session *calling_session,
+			     uint32_t param_types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INOUT,
+						TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+
+	const struct user_ta_ctx *utc = to_user_ta_ctx(calling_session->ctx);
+
+	TEE_Result res = TEE_SUCCESS;
+
+	uint32_t ocall_id = 0;
+	uint32_t ocall_pt = 0;
+	struct utee_params *ocall_up = NULL;
+	size_t ocall_up_size = 0;
+
+	struct mobj *mobj = NULL;
+	size_t mobj_size = 0;
+	void *mobj_va = 0;
+
+	struct thread_param rpc_params[THREAD_RPC_MAX_NUM_PARAMS] = { 0 };
+	const size_t rpc_num_params = ARRAY_SIZE(rpc_params);
+
+	if (param_types != exp_pt)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ocall_up_size = params[1].memref.size;
+	if (ocall_up_size != sizeof(*ocall_up))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ocall_id = params[0].value.a;
+	ocall_pt = params[0].value.b;
+	ocall_up = (struct utee_params *)params[1].memref.buffer;  // User ptr
+
+	res = check_access_rights(&utc->uctx, ocall_up);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = compute_mobj_size(ocall_up, &mobj_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	if (mobj_size) {
+		mobj = thread_rpc_alloc_client_app_payload(mobj_size);
+		if (!mobj)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		mobj_va = mobj_get_va(mobj, 0);
+		if (!mobj_va) {
+			res = TEE_ERROR_GENERIC;
+			goto exit;
+		}
+	}
+
+	rpc_params[0] = THREAD_PARAM_VALUE(INOUT, ocall_id, ocall_pt, 0);
+	rpc_params[1] = THREAD_PARAM_VALUE(IN, 0, 0, 0);
+	tee_uuid_to_octets((uint8_t *)&rpc_params[1].u.value,
+			   &session->clnt_id.uuid);
+
+	res = pre_process(rpc_params + 2, ocall_up, mobj, mobj_va);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = thread_rpc_cmd(OPTEE_RPC_CMD_OCALL, rpc_num_params, rpc_params);
+	if (res != TEE_SUCCESS) {
+		mobj_put(mobj);
```

Comment:
why `mobj_put()` here and `thread_rpc_free_client_app_payload()` on exit?

Reply:
I added a comment to explain the rationale. Do you see a different possible behavior?

Reply:
Modified as suggested.